src/Symfony/Component/Finder/Iterator/VcsIgnoredFilterIterator.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Finder\Iterator;
+
+use Iterator;
+use Symfony\Component\Finder\Gitignore;
+
+class VcsIgnoredFilterIterator extends \FilterIterator
+{
+    /**
+     * @var string
+     */
+    private $baseDir;
+
+    public function __construct(Iterator $iterator, string $baseDir)
+    {
+        $this->baseDir = $baseDir;
+
+        parent::__construct($iterator);
+    }
+
+    public function accept(): bool
+    {
+        $file = $this->current();
+
+        $fileRealPath = $file->getRealPath();
+        if ($file->isDir()) {
+            $fileRealPath .= '/';
+        }
+
+        $parentDirectory = $fileRealPath;
+
+        do {
+            $parentDirectory = \dirname($parentDirectory);
+            $gitignoreFilePath = "{$parentDirectory}/.gitignore";
+
+            if (!file_exists($gitignoreFilePath)) {
+                continue;
+            }
+
+            if (!is_file($gitignoreFilePath) || !is_readable($gitignoreFilePath)) {
+                throw new \RuntimeException("The \"ignoreVCSIgnored\" option cannot be used by the Finder as the \"{$gitignoreFilePath}\" file is not readable.");
+            }
+
+            $relativeFilePath = substr($fileRealPath, \strlen($parentDirectory) + 1);
+
+            $regex = Gitignore::toRegex(file_get_contents($gitignoreFilePath));
```

Comment:
This can probably be optimized by caching the regular expressions to avoid reading the files multiple times.